### PR TITLE
Make opnorm( 1x0 matrix ) zero

### LINKED ISSUE #835: opnorm fails on an 1 x 0 matrix
The following computation leads to an error:
```
julia> a = rand(1,1)
1×1 Matrix{Float64}:
 0.86747552215521

julia> t = a[:,2:1]
1×0 Matrix{Float64}

julia> opnorm(t)
ERROR: TypeError: in typeassert, expected Tuple, got a value of type Nothing
Stacktrace:
 [1] generic_normInf
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:465 [inlined]
 [2] normInf
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:556 [inlined]
 [3] generic_norm2(x::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:497
 [4] norm2(x::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\dense.jl:105
 [5] opnorm2
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:705 [inlined]
 [6] opnorm(A::Matrix{Float64}, p::Int64)
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:767
 [7] opnorm(A::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:766
 [8] top-level scope
   @ REPL[192]:1
```

But, there is no error if the matrix has a slightly different shape:
```
julia> a = rand(2,1);

julia> t = a[:,2:1]
2×0 Matrix{Float64}

julia> opnorm(t)
0.0
```

### LINKED ISSUE #40370: opnorm fails on an 1 x 0 matrix
The following computation leads to an error:
```
julia> a = rand(1,1)
1×1 Matrix{Float64}:
 0.86747552215521

julia> t = a[:,2:1]
1×0 Matrix{Float64}

julia> opnorm(t)
ERROR: TypeError: in typeassert, expected Tuple, got a value of type Nothing
Stacktrace:
 [1] generic_normInf
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:465 [inlined]
 [2] normInf
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:556 [inlined]
 [3] generic_norm2(x::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:497
 [4] norm2(x::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\dense.jl:105
 [5] opnorm2
   @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:705 [inlined]
 [6] opnorm(A::Matrix{Float64}, p::Int64)
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:767
 [7] opnorm(A::Matrix{Float64})
   @ LinearAlgebra C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\generic.jl:766
 [8] top-level scope
   @ REPL[192]:1
```

But, there is no error if the matrix has a slightly different shape:
```
julia> a = rand(2,1);

julia> t = a[:,2:1]
2×0 Matrix{Float64}

julia> opnorm(t)
0.0
```

### PR DESCRIPTION
**Pull request by ghost**
Fixes #40370.

### CODE REVIEW

**Review by oscardssmith — APPROVED** [review]

**Comment by antoine-levitt** [comment]
There has been discussion of a related issue at https://github.com/JuliaLang/julia/pull/38372, where it's still not clear to me what the exact semantics of empty matrices should be. This opnorm change seems much less ambiguous though, so +1.

File: stdlib/LinearAlgebra/src/generic.jl [hunk 1]
```diff
@@ -702,9 +702,10 @@ end
 function opnorm2(A::AbstractMatrix{T}) where T
     require_one_based_indexing(A)
     m,n = size(A)
-    if m == 1 || n == 1 return norm2(A) end
     Tnorm = typeof(float(real(zero(T))))
-    (m == 0 || n == 0) ? zero(Tnorm) : convert(Tnorm, svdvals(A)[1])
+    if m == 0 || n == 0 return zero(Tnorm) end
+    if m == 1 || n == 1 return norm2(A) end
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/generic.jl:707]: Do we want a `convert(Tnorm, norm2(A))` just to be paranoid about type-stability?  (I'm not sure why it's required for the `svdvals` either.)

File: stdlib/LinearAlgebra/test/generic.jl [hunk 2]
```diff
@@ -209,6 +209,9 @@ end
 @test norm([2.4e-322, 4.4e-323], 3) ≈ 2.4e-322
 @test_throws ArgumentError opnorm(Matrix{Float64}(undef,5,5),5)
 
+# operator norm for zero-dimensional domain is zero (see #40370)
+@test opnorm(fill(1,1,1)[:,2:1]) == 0.0
```
Comment (stevengj) [inline stdlib/LinearAlgebra/test/generic.jl:213]: Probably should test both `m == 0` and `n == 0`.

Also, if you are special-casing the `m == 1 || n == 1` case, those should both be tested.

**Review by dkarrasch — APPROVED** [review]

File: stdlib/LinearAlgebra/src/generic.jl (hunk 1 above)
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:707]: I've removed the convert on the svdvals and added inference tests.

**Pull request merged**